src/CachingIterator.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace MongoDB;
+
+class CachingIterator implements \Iterator, \Countable
+{
+    /**
+     * @var \Traversable
+     */
+    private $iterator;
+
+    /**
+     * @var array
+     */
+    private $items = [];
+
+    /**
+     * @var bool
+     */
+    private $iteratorExhausted = false;
+
+    /**
+     * @param \Traversable $iterator
+     */
+    public function __construct(\Traversable $iterator)
+    {
+        $this->iterator = $this->wrapTraversable($iterator);
+        $this->storeCurrentItem();
+    }
+
+    /**
+     * @return int
+     */
+    public function count()
+    {
+        $this->exhaustIterator();
+        return count($this->items);
+    }
+
+    /**
+     * @return mixed
+     */
+    public function current()
+    {
+        return current($this->items);
+    }
+
+    /**
+     * @return mixed
+     */
+    public function key()
+    {
+        return key($this->items);
+    }
+
+    /**
+     * @return void
+     */
+    public function next()
+    {
+        if (! $this->iteratorExhausted) {
+            $this->iterator->next();
+            $this->storeCurrentItem();
+        }
+
+        next($this->items);
+    }
+
+    /**
+     * @return void
+     */
+    public function rewind()
+    {
+        $this->exhaustIterator();
+        reset($this->items);
+    }
+
+    /**
+     * @return bool
+     */
+    public function valid()
+    {
+        return $this->key() !== null;
+    }
+
+    /**
+     * Ensures the original iterator is fully consumed and all items cached
+     */
+    private function exhaustIterator()
+    {
+        while (!$this->iteratorExhausted) {
+            $this->next();
+        }
+    }
+
+    /**
+     * Stores the current item
+     */
+    private function storeCurrentItem()
+    {
+        if (null === $key = $this->iterator->key()) {
+            return;
+        }
+
+        $this->items[$key] = $this->iterator->current();
+    }
+
+    /**
+     * @param \Traversable $traversable
+     * @return \Generator
+     */
+    private function wrapTraversable(\Traversable $traversable)
+    {
+        foreach ($traversable as $key => $value) {
+            yield $key => $value;
```

Comment:
Conflicts with [`composer.json` requiring PHP >= 5.4](https://github.com/mongodb/mongo-php-library/blob/318d4b95438a2ea68a550c2878678097f63fb9fe/composer.json#L13). I'll see what we can do about bumping this for 1.2.0.